### ISSUE TITLE
Datalog: 5 million fact challenge

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#1893]
Flix has no problem allocating a 50 million element list:

```
pub def main(_: Array[String]): Int32 & Impure =
    let l = List.range(0, 50_000_000);
    l |> List.length |> println;
    0
```

using less than 4GB memory.

I'd like to see us support a Datalog program that can process and store ~5 million facts.

**Comment by bdahse** [comment]
Any specific program you have in mind? We're not going to compute the transitive closure anytime soon, that's for sure :) (and certainly not in less than 4gb memory).

**Comment by magnus-madsen** [comment]
We could proceed as follows:

1. Project 5M facts into a relation.
2. Project 5M facts out of a relation.
3. Copy 1M facts from one relation into another (using a single rule).
4. Compute the transitive closure of a connected graph with 5M edges total. (i.e. if |edges|=N^2 then we start with 1000 edges).
5. Compute the transitive closure of a fully connected graph with 5M edges totao (i.e. if |edges|=N^3 then we start with 170 edges)

**Comment by magnus-madsen** [comment]
For (1) and (2), note that:

```
Project (Int, Int) into A (n = 1048576)                    2        940401 us/op
Select (Int, Int) from A (n = 4096)                        1       1023867 us/op
```

which shows that (1) is arguably achieved 1M facts/second. Whereas for (2) the throughput is around 4K/second.

Presumably because the translation to RAM is N^2.

**Comment by bdahse** [comment]
> 
> 
> We could proceed as follows:
> 
>     1. Project 5M facts into a relation.
> 
>     2. Project 5M facts out of a relation.
> 
>     3. Copy 1M facts from one relation into another (using a single rule).
> 
>     4. Compute the transitive closure of a connected graph with 5M edges total. (i.e. if |edges|=N^2 then we start with 1000 edges).
> 
>     5. Compute the transitive closure of a fully connected graph with 5M edges totao (i.e. if |edges|=N^3 then we start with 170 edges)

For (2) what is the difference between `solve m project A`  and `query m select (x, y) from A(x, y)` ? The latter copies the `A` relation into the `$Result` relation before projection? If so isn't the difference between the two the same as (2) vs (3)?

**Comment by magnus-madsen** [comment]
I think (3) is redundant. But certainly solve and query are different.

**Comment by bdahse** [comment]
> 
> 
> I think (3) is redundant. But certainly solve and query are different.

Right, I'm just trying to understand the difference. So `solve m project A` ends with a call to `Fixpoint.project` in Solver.flix and you stay in Datalog-land.  `query m select (x, y) from A(x, y)` ends with a call to `Fixpoint.facts2` and you're back in Flix?

**Comment by magnus-madsen** [comment]
> 
> 
> > I think (3) is redundant. But certainly solve and query are different.
> 
> Right, I'm just trying to understand the difference. So `solve m project A` ends with a call to `Fixpoint.project` in Solver.flix and you stay in Datalog-land. `query m select (x, y) from A(x, y)` ends with a call to `Fixpoint.facts2` and you're back in Flix?

Thats right.

**Comment by bdahse** [comment]
> 
> 
> > > I think (3) is redundant. But certainly solve and query are different.
> > 
> > 
> > Right, I'm just trying to understand the difference. So `solve m project A` ends with a call to `Fixpoint.project` in Solver.flix and you stay in Datalog-land. `query m select (x, y) from A(x, y)` ends with a call to `Fixpoint.facts2` and you're back in Flix?
> 
> Thats right.

Right now even a million facts is infeasible for (2). I suppose it will stay infeasible until the compiler is no longer N^2. Once we can at least get a measurement, I suggest optimizing `Array.filter` like we talked about and and `factsX` by getting rid of an additional `Array.map` for every fact.

**Comment by magnus-madsen** [comment]
I can insert up to 8 million facts, no problem. It seems to scale well, as long as there is enough memory.

**Comment by bdahse** [comment]
Can you project out 8 million facts? The million I was talking about is the `Select (Int, Int) from A (n = 2 ** 20)` benchmark.

**Comment by magnus-madsen** [comment]
> 
> 
> Can you project out 8 million facts? The million I was talking about is the `Select (Int, Int) from A (n = 2 ** 20)` benchmark.

No? Not even close. I mean, we did not change anything in that respect yet?

**Comment by magnus-madsen** [comment]
@bdahse 

Here is a proposal for an updated list of goals:

Assume we have 4GB of memory for the JVM:

- [x] Project 4M facts into a relation (project into) (in seconds.)
- [ ] Project 4M facts out of a relation (select from) (in seconds.)
- [ ] Compute the cross product of two 2K relations (i.e. 4M facts) (in tens of seconds.)
- [ ] Compute the cross product of three 128 relations (i.e. 2M facts) (in tens of seconds.)
- [ ] Compute the intersection of A and B where A and B have 2M facts each and their overlap is 1M facts (in tens of seconds.)
- [ ] Compute the difference of A and B where A has 4M facts and B has 2M facts (in tens of seconds.)
- [ ] Compute the transitive closure of a linear chain with 2K edges (i.e. ~2M facts) (in tens of seconds.)
- [ ] Compile 1K rules. (in seconds.)

I can try to implement benchmarks for these.

**Comment by magnus-madsen** [comment]
I added benchmarks for most of these. These are my current results:

```
Benchmark                                         Iterations          Time (avg)
Project (Int, Int) into A (n = 16)                     35366            10 us/op
Project (Int, Int) into A (n = 256)                    14595           115 us/op
Project (Int, Int) into A (n = 4096)                    1054          1861 us/op
Project (Int, Int) into A (n = 65536)                     69         30472 us/op
Project (Int, Int) into A (n = 1048576)                    4        687484 us/op
Select (Int, Int) from A (n = 16)                        382           644 us/op
Select (Int, Int) from A (n = 256)                        77          9506 us/op
Select (Int, Int) from A (n = 4096)                        3       1017669 us/op
Solve (Closure) (n = 64)                                 585          1143 us/op
Solve (Closure) (n = 256)                                223          3135 us/op
Solve (Closure) (n = 1024)                                71         16825 us/op
Solve (Closure) (n = 4096)                                11        116201 us/op
Solve (Closure) (n = 16384)                                4        846384 us/op
Solve (Cross Product 2) (n = 256)                        954          1734 us/op
Solve (Cross Product 2) (n = 1024)                       395          6722 us/op
Solve (Cross Product 2) (n = 4096)                        98         30032 us/op
Solve (Cross Product 2) (n = 16384)                       28        136484 us/op
Solve (Cross Product 2) (n = 65536)                        6        604418 us/op
Solve (Cross Product 2) (n = 262144)                       1       2683599 us/op
Solve (Cross Product 2) (n = 1048576)                      1      11915812 us/op
Solve (Cross Product 3) (n = 4096)                       135         27891 us/op
Solve (Cross Product 3) (n = 32768)                       14        295552 us/op
Solve (Cross Product 3) (n = 262144)                       1       2848670 us/op
Solve (Intersect) (n = 16)                              1642           524 us/op
Solve (Intersect) (n = 256)                              159         20327 us/op
Solve (Intersect) (n = 4096)                               1       3777140 us/op
Compile Rules (n = 16)                                   132          3954 us/op
Compile Rules (n = 256)                                   12        105810 us/op
Compile Rules (n = 4096)                                   1      13132843 us/op
```

**Comment by bdahse** [comment]
Thanks, I have implemented benchmarks for selection and projection that I'm using so far. I plan to submit a benchmark that gets rid of the infamous `Monoid[RamStmt]` today. Hopefully then we will be able to progress past the `Select` benchmark.

**Comment by bdahse** [comment]
These are my latest results with #1911 :
```
Benchmark                                         Iterations          Time (avg)
Project (Int, Int) into A (n = 16)                     97713            12 us/op
Project (Int, Int) into A (n = 256)                    97024           131 us/op
Project (Int, Int) into A (n = 4096)                    9540          2056 us/op
Project (Int, Int) into A (n = 65536)                    352         42804 us/op
Project (Int, Int) into A (n = 1048576)                   21       1214842 us/op
Select (Int, Int) from A (n = 16)                       1355           883 us/op
Select (Int, Int) from A (n = 256)                       165          5133 us/op
Select (Int, Int) from A (n = 4096)                       28        108162 us/op
Select (Int, Int) from A (n = 65536)                       6       1852285 us/op
Solve (Closure) (n = 64)                                 747          1827 us/op
Solve (Closure) (n = 256)                                467          4695 us/op
Solve (Closure) (n = 1024)                                94         30701 us/op
Solve (Closure) (n = 4096)                                29        157762 us/op
Solve (Closure) (n = 16384)                               14       1260763 us/op
Solve (Cross Product 2) (n = 256)                       2928          2352 us/op
Solve (Cross Product 2) (n = 1024)                      1143          9157 us/op
Solve (Cross Product 2) (n = 4096)                       357         39068 us/op
Solve (Cross Product 2) (n = 16384)                       83        169690 us/op
Solve (Cross Product 2) (n = 65536)                       22        762625 us/op
Solve (Cross Product 2) (n = 262144)                       5       3515029 us/op
Solve (Cross Product 2) (n = 1048576)                      1      15171116 us/op
Solve (Cross Product 3) (n = 4096)                       301         38621 us/op
Solve (Cross Product 3) (n = 32768)                       34        373378 us/op
Solve (Cross Product 3) (n = 262144)                       4       3704068 us/op
Solve (Intersect) (n = 16)                              4370           766 us/op
Solve (Intersect) (n = 256)                             1176          7970 us/op
Solve (Intersect) (n = 4096)                              71        156459 us/op
Compile Rules (n = 16)                                   388          5351 us/op
Compile Rules (n = 256)                                   34         88821 us/op
Compile Rules (n = 4096)                                   5       1802612 us/op
```

Notice the `Select (Int, Int) from A (n = 4096)` benchmark. Compared to my results in https://github.com/flix/flix/issues/1878#issuecomment-826679881 , that's a 11.5x improvement!

Projecting out is still a lot slower than projecting in. That's inevitable given that projecting out entails many more steps: 1) compilation to RAM 2) copying into a red-black tree 3) flattening the red-black tree into an array of facts 4) converting array of facts to array of tuples. Note that there are multiple passes over the RAM program in (1). Each pass copies all the facts. I.e. no transformation is done to them.

**Comment by magnus-madsen** [comment]
Can you scale some of the other benchmarks to higher n? What about select? Can it go to higher n?

**Comment by bdahse** [comment]
> Can you scale some of the other benchmarks to higher n? What about select? Can it go to higher n?

Select can go to 65K now, but not 1M. Although it seems to scale alright. From N=4096 to N=65536 is a factor 16 but the running time increased by 17.1x.

This is how high I can scale the benchmarks before I run out of memory or it just takes too long (a VM with just 6GB RAM):
```
Benchmark                                         Iterations          Time (avg)
Project (Int, Int) into A (n = 16)                    120181            15 us/op
Project (Int, Int) into A (n = 256)                    81032           130 us/op
Project (Int, Int) into A (n = 4096)                    9813          2204 us/op
Project (Int, Int) into A (n = 65536)                    545         36195 us/op
Project (Int, Int) into A (n = 1048576)                   23       1371998 us/op
Solve (Closure) (n = 64)                                 639          7997 us/op
Solve (Closure) (n = 256)                                294         13346 us/op
Solve (Closure) (n = 1024)                                75        256029 us/op
Solve (Closure) (n = 4096)                                31        192261 us/op
Solve (Closure) (n = 16384)                               16       1281512 us/op
Solve (Closure) (n = 65536)                                2       9662301 us/op
Solve (Closure) (n = 262144)                               1      57857605 us/op
Solve (Cross Product 2) (n = 256)                       2842          2426 us/op
Solve (Cross Product 2) (n = 1024)                      1038          9122 us/op
Solve (Cross Product 2) (n = 4096)                       300         38711 us/op
Solve (Cross Product 2) (n = 16384)                      104        171258 us/op
Solve (Cross Product 2) (n = 65536)                       25        773929 us/op
Solve (Cross Product 2) (n = 262144)                       5       3505629 us/op
Solve (Cross Product 2) (n = 1048576)                      1      16656975 us/op
Solve (Cross Product 3) (n = 4096)                       416         36350 us/op
Solve (Cross Product 3) (n = 32768)                       29        375725 us/op
Solve (Cross Product 3) (n = 262144)                       4       3802546 us/op
Solve (Intersect) (n = 16)                              3405           772 us/op
Solve (Intersect) (n = 256)                             1068          8052 us/op
Solve (Intersect) (n = 4096)                              62        154043 us/op
Solve (Intersect) (n = 65536)                              6       3090487 us/op
Compile Rules (n = 16)                                   337          6221 us/op
Compile Rules (n = 256)                                   36         97970 us/op
Compile Rules (n = 4096)                                   7       1557063 us/op
```

We're still quite far behind the goal. Computing the transitive closure of 262K facts (or is it 131K?) takes 57 seconds. That's a factor 10-20 off the goal of 2M facts.

**Comment by magnus-madsen** [comment]
It seems select is still comparatively slow compared to say cross product or intersection? (Am I understanding that right?)

**Comment by bdahse** [comment]
> 
> 
> It seems select is still comparatively slow compared to say cross product or intersection? (Am I understanding that right?)

Intersection seems slower to me: 3.5 secs vs 1.8 secs for N = 65K ?
But for cross product you seem to be right (0.7 secs vs 1.8 secs for N = 65K). I suspect that is because a cross product of size 65K only requires two relations of 256 facts each. To me this indicates that computing 65K facts is "fast" but dealing with 65K facts is slow.

**Comment by bdahse** [comment]
I implemented the simplifier as described in #1895 . It seems to be amazing for transitive closure, but it slightly hurts benchmarks that are bottlenecked by the number of facts because it is an additional phase. I have an idea for how to fix that though, which should also be a general improvement. The exact results are as follows:
```
Benchmark                                         Iterations          Time (avg)
Project (Int, Int) into A (n = 16)                     81877            11 us/op
Project (Int, Int) into A (n = 256)                    81289           141 us/op
Project (Int, Int) into A (n = 4096)                    8518          1987 us/op
Project (Int, Int) into A (n = 65536)                    552         32954 us/op
Project (Int, Int) into A (n = 1048576)                   31       1221969 us/op
Select (Int, Int) from A (n = 16)                       1414          1416 us/op
Select (Int, Int) from A (n = 256)                       207          7737 us/op
Select (Int, Int) from A (n = 4096)                       38         89912 us/op
Select (Int, Int) from A (n = 65536)                       5       1981714 us/op
Solve (Closure) (n = 64)                                1428          1929 us/op
Solve (Closure) (n = 256)                                724          3495 us/op
Solve (Closure) (n = 1024)                               209          9745 us/op
Solve (Closure) (n = 4096)                                52         39734 us/op
Solve (Closure) (n = 16384)                               23        201597 us/op
Solve (Closure) (n = 65536)                               12        860753 us/op
Solve (Closure) (n = 262144)                               4       4038372 us/op
Solve (Cross Product 2) (n = 256)                       3215          2004 us/op
Solve (Cross Product 2) (n = 1024)                      1099          8023 us/op
Solve (Cross Product 2) (n = 4096)                       276         40731 us/op
Solve (Cross Product 2) (n = 16384)                       93        160456 us/op
Solve (Cross Product 2) (n = 65536)                       23        773265 us/op
Solve (Cross Product 2) (n = 262144)                       5       3755776 us/op
Solve (Cross Product 2) (n = 1048576)                      1      14591156 us/op
Solve (Cross Product 3) (n = 4096)                       502         36379 us/op
Solve (Cross Product 3) (n = 32768)                       50        375737 us/op
Solve (Cross Product 3) (n = 262144)                       4       4087558 us/op
Solve (Intersect) (n = 16)                              5398           604 us/op
Solve (Intersect) (n = 256)                             1200          7395 us/op
Solve (Intersect) (n = 4096)                              70        147456 us/op
Solve (Intersect) (n = 65536)                              5       3337457 us/op
Solve (Intersect) (n = 1048576)                            1      70091508 us/op
Compile Rules (n = 16)                                   347          7829 us/op
Compile Rules (n = 256)                                   34         82911 us/op
Compile Rules (n = 4096)                                   5       1448666 us/op
```
Closure of size 262K just dropped from 57.8 seconds to 4.0 seconds! 14.3x :+1:
